utils/metrics_batch_builder.go

```diff
@@ -0,0 +1,333 @@
+			Samples: optsm.Samples,
+		}
+		if optsm.Label < uint32(int(livekit.MetricLabel_METRIC_LABEL_PREDEFINED_MAX_VALUE)) {
+			participantIdentity, ok := getStrDataForIndex(other, optsm.ParticipantIdentity)
+			if ok && m.isLabelFiltered(livekit.MetricLabel(optsm.Label), livekit.ParticipantIdentity(participantIdentity)) {
+				continue
+			}
+
+			ptsm.Label = optsm.Label
+		} else {
+			if tidx, ok := m.translateStrDataIndex(other.StrData, optsm.Label); ok {
+				ptsm.Label = tidx
+			}
+		}
+
+		if tidx, ok := m.translateStrDataIndex(other.StrData, optsm.ParticipantIdentity); ok {
+			ptsm.ParticipantIdentity = tidx
+		}
+
+		if tidx, ok := m.translateStrDataIndex(other.StrData, optsm.TrackSid); ok {
+			ptsm.TrackSid = tidx
+		}
+
+		if tidx, ok := m.translateStrDataIndex(other.StrData, optsm.Rid); ok {
+			ptsm.Rid = tidx
+		}
+
+		m.MetricsBatch.TimeSeries = append(m.MetricsBatch.TimeSeries, ptsm)
+	}
+
+	for _, opem := range other.Events {
+		pem := &livekit.EventMetric{}
+		if opem.Label < uint32(int(livekit.MetricLabel_METRIC_LABEL_PREDEFINED_MAX_VALUE)) {
+			participantIdentity, ok := getStrDataForIndex(other, opem.ParticipantIdentity)
+			if ok && m.isLabelFiltered(livekit.MetricLabel(opem.Label), livekit.ParticipantIdentity(participantIdentity)) {
+				continue
+			}
+
+			pem.Label = opem.Label
+		} else {
+			if tidx, ok := m.translateStrDataIndex(other.StrData, opem.Label); ok {
+				pem.Label = tidx
+			}
+		}
+
+		if tidx, ok := m.translateStrDataIndex(other.StrData, opem.ParticipantIdentity); ok {
+			pem.ParticipantIdentity = tidx
+		}
+
+		if tidx, ok := m.translateStrDataIndex(other.StrData, opem.TrackSid); ok {
+			pem.TrackSid = tidx
+		}
+
+		pem.StartTimestampMs = opem.StartTimestampMs
+		pem.EndTimestampMs = opem.EndTimestampMs
+		pem.NormalizedStartTimestamp = opem.NormalizedStartTimestamp
+		pem.NormalizedEndTimestamp = opem.NormalizedEndTimestamp
+
+		pem.Metadata = opem.Metadata
+
+		if tidx, ok := m.translateStrDataIndex(other.StrData, opem.Rid); ok {
+			pem.Rid = tidx
+		}
+
+		m.MetricsBatch.Events = append(m.MetricsBatch.Events, pem)
+	}
+}
+
+func (m *MetricsBatchBuilder) IsEmpty() bool {
+	return len(m.MetricsBatch.TimeSeries) == 0 && len(m.MetricsBatch.Events) == 0
+}
+
+func (m *MetricsBatchBuilder) isLabelFiltered(label livekit.MetricLabel, participantIdentity livekit.ParticipantIdentity) bool {
+	if participantIdentity == m.restrictedLabels.ParticipantIdentity {
+		// all labels allowed for restricted participant
+		return false
+	}
+
+	for _, mlr := range m.restrictedLabels.LabelRanges {
+		if label >= mlr.StartInclusive && label <= mlr.EndInclusive {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (m *MetricsBatchBuilder) getStrDataIndex(s string) uint32 {
+	idx, ok := m.stringData[s]
+	if !ok {
+		m.MetricsBatch.StrData = append(m.MetricsBatch.StrData, s)
+		idx = uint32(int(livekit.MetricLabel_METRIC_LABEL_PREDEFINED_MAX_VALUE) + len(m.MetricsBatch.StrData) - 1)
```

Comment:
Using  the pre-defined offset of 4096 for all entities pointing to `StrData`. Technically, the offset is needed only for the label. If we do that, for the other fields that index into `StrData` (like `ParticipantIdentity` etc.) should either
- become `optional` so that `0` value can be distinguished from `not present` OR
- use offset of `+1` to indicate index into `StrData` and `0` can be interpreted as `not present`.

Tagging @davidliu for any changes needed in Android if we decide to change this. I told him when he was working on this several weeks back to do `4096` as base offset for all.